base/array.jl

```diff
@@ -981,3 +981,18 @@ end
 symdiff(a) = a
 symdiff(a, b) = union(setdiff(a,b), setdiff(b,a))
 symdiff(a, b, rest...) = symdiff(a, symdiff(b, rest...))
+
+# Lightweight wrapper type for values(::Array)
+immutable ArrayIterator{T,N}
+    array::Array{T,N}
+end
+start(ai::ArrayIterator) = start(ai.array)
+next(ai::ArrayIterator,k) = next(ai.array,k)
+done(ai::ArrayIterator,k) = done(ai.array,k)
+
+length(v::ArrayIterator) = length(v.array)
+isempty(v::ArrayIterator) = isempty(v.array)
+eltype{T,N}(::Type{ArrayIterator{T,N}}) = T
+
+keys(x::Array) = CartesianRange(size(x))
+values(x::Array) = ArrayIterator(x)
```

Comment:
If we have this at all, I would prefer `values(x::AbstractArray) = x`; if the main purpose is to use this as an iterator, then I don't understand why you would need to prevent mutation.   `for x in values(a)` or `for x in a` already doesn't allow mutation (unless `x` is mutable, of course).